### ISSUE TITLE
[Optimization] SDPA Fast Path + Cached Bool Window Mask

### ISSUE THREAD

**Issue by z-a-f** [maida-ai/Research#34]
### Summary
Speed up the attention path by using SDPA's causal fast path when $W \ge T$, and a **cached boolean band mask** when $W < T$. Remove the additive `-inf` mask. Standardize dropout handling.

### Motivation
Avoids mask re-build cost, unlocks fused SDPA/FlashAttention kernels, and reduces dtype pitfalls.

### Scope

* Replace additive mask with cached **bool** mask shaped $(1,1,T,T)$, where **True = allowed**.
* Use `is_causal=True` only when **no mask** is passed.
* Standardize dropout probability: `p = self.drop.p if self.training else 0.0`.

### Implementation Sketch

```python
# __init__
self._mask_cache = {}

def _get_window_mask(self, T: int, W: int, device) -> torch.Tensor:
    key = (T, W, device)
    if key in self._mask_cache:
        return self._mask_cache[key]
    i = torch.arange(T, device=device)[:, None]
    j = torch.arange(T, device=device)[None, :]
    band = (j <= i) & ((i - j) < W)  # causal + left window
    mask = band.view(1,1,T,T)        # bool mask
    self._mask_cache[key] = mask
    return mask

# in _compute_attention:
if self.window_size >= T:
    attn_mask, is_causal = None, True
else:
    attn_mask = self._get_window_mask(T, self.window_size, x.device)
    is_causal = False

out = F.scaled_dot_product_attention(Q, K, V, attn_mask=attn_mask,
                                     dropout_p=self.drop.p if self.training else 0.0,
                                     is_causal=is_causal)
```

### Tests

* Windowed correctness: tokens outside window do not affect target.
* Causal fast path selected when $W \ge T$.
* No dtype mismatch or warning logs.

### Acceptance Criteria

* [ ] Bool mask cached & reused; no additive masks kept.
* [ ] Throughput improves vs. baseline on $T=512$, $W=256$.
* [ ] Unit tests for windowing/causality pass.

### General Notes

* Keep changes **behind flags** where practical (`impl`, `proj_stride`, `num_kv_heads`, `overlap_streams`, `compile`), so we can A/B easily.
* Add **unit tests** for shapes, continuity (state carry), and streaming equivalence.
* Add at least one **micro-benchmark** result in the PR description (before/after tokens/sec on T=512 or 1024).
